[jit] Allow 'Any'  to appear as a type argument.

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26788 [jit] clean up exported source format
* #26767 [jit] Remove attempToRecoverType
* **#26572 [jit] Allow 'Any'  to appear as a type argument.**

Combined with isinstance specialization this allows a degree of polymorphic
functions to work without needing to use our weirder overload hacks.

We do not define any operators on Any, so the only thing you can do with it
is to put it in containers or type refine it using an isinstance check.
Any is restricted from appearing in non-argument position because we
cannot restore type tags if it ends up as a field in a class.

Differential Revision: [D17530643](https://our.internmc.facebook.com/intern/diff/D17530643)

suo commented on test/test_jit.py:
```diff
@@ -6801,6 +6801,24 @@ def foo2(a : Optional[int], b : Optional[int]):
+        self.assertEqual(any_refinement(3, 4), 7)
+        self.assertEqual(any_refinement(3, "hi"), 0)
+
+    def test_any_in_class_fails(self):
```
Can we add tests that verify that `Any` is not allowed in Tuples, NamedTuples, or Interfaces as well?

suo commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -782,6 +782,19 @@ struct CAFFE2_API NamedType : public Type {
+// static types in named types to reconstruct type tags of loaded
+// values. Lifting this restriction requires solving the serialization
+// problem first.
+CAFFE2_API void checkNoAny(
```
I am a little wary of having this enforcement as opt-in. For example, if someone goes in and adds attributes to interfaces and forgets to include this check, we've very quickly produced the bug this comment describes.

zdevito replied:
The specific example of interfaces will not actually cause this bug, but I agree in spirit. But how would we make it more likely to be called? There isn't a more common place to put it.

suo reviewed (approved):
ngnt

zdevito committed:
Update on "[jit] Allow 'Any'  to appear as a type argument."

[jit] Allow 'Any'  to appear as a type argument.

Combined with isinstance specialization this allows a degree of polymorphic
functions to work without needing to use our weirder overload hacks.

We do not define any operators on Any, so the only thing you can do with it
is to put it in containers or type refine it using an isinstance check.
Any is restricted from appearing in non-argument position because we
cannot restore type tags if it ends up as a field in a class.

gh-metadata: pytorch pytorch 26572 gh/zdevito/113/head

zdevito committed:
Update on "[jit] Allow 'Any'  to appear as a type argument."

[jit] Allow 'Any'  to appear as a type argument.

Combined with isinstance specialization this allows a degree of polymorphic
functions to work without needing to use our weirder overload hacks.

We do not define any operators on Any, so the only thing you can do with it
is to put it in containers or type refine it using an isinstance check.
Any is restricted from appearing in non-argument position because we
cannot restore type tags if it ends up as a field in a class.

gh-metadata: pytorch pytorch 26572 gh/zdevito/113/head

zdevito committed:
Update on "[jit] Allow 'Any'  to appear as a type argument."

[jit] Allow 'Any'  to appear as a type argument.

Combined with isinstance specialization this allows a degree of polymorphic
functions to work without needing to use our weirder overload hacks.

We do not define any operators on Any, so the only thing you can do with it
is to put it in containers or type refine it using an isinstance check.
Any is restricted from appearing in non-argument position because we
cannot restore type tags if it ends up as a field in a class.

gh-metadata: pytorch pytorch 26572 gh/zdevito/113/head

zdevito committed:
Update on "[jit] Allow 'Any'  to appear as a type argument."

[jit] Allow 'Any'  to appear as a type argument.

Combined with isinstance specialization this allows a degree of polymorphic
functions to work without needing to use our weirder overload hacks.

We do not define any operators on Any, so the only thing you can do with it
is to put it in containers or type refine it using an isinstance check.
Any is restricted from appearing in non-argument position because we
cannot restore type tags if it ends up as a field in a class.

gh-metadata: pytorch pytorch 26572 gh/zdevito/113/head

zdevito committed:
Update on "[jit] Allow 'Any'  to appear as a type argument."

[jit] Allow 'Any'  to appear as a type argument.

Combined with isinstance specialization this allows a degree of polymorphic
functions to work without needing to use our weirder overload hacks.

We do not define any operators on Any, so the only thing you can do with it
is to put it in containers or type refine it using an isinstance check.
Any is restricted from appearing in non-argument position because we
cannot restore type tags if it ends up as a field in a class.

gh-metadata: pytorch pytorch 26572 gh/zdevito/113/head

tangleintel commented on torch/csrc/jit/pybind_utils.h:
```diff
@@ -287,37 +285,24 @@ inline bool isTraceableType(TypePtr type) {
+inline Stack toTraceableStack(const py::tuple& inputs) {
+  auto info = toTypeInferredIValue(inputs);
+  AT_CHECK(
+      isTraceableType(info.type()),
```
Hi, @zdevito 
May I know why we only support Tensors and (possibly nested) Lists, Dicts, and Tuples of Tensors?
What if we passed a `None` type of python, and what will happen that makes us not support it?
